discv5/discv5-theory.md

```diff
@@ -41,6 +40,141 @@ IP address and port.
 If the endpoint cannot be determined (e.g. when the NAT doesn't support 'full-cone'
 translation), implementations should omit IP address and UDP port from the record.
 
+## Sessions
+
+Discovery communication is encrypted and authenticated using session keys, established in
+the handshake.
+
+### Handshake
+
+Since every node participating in the network acts as both client and server, a handshake
+can be initiated by either side of communication at any time. In the following
+definitions, we assume that node A wishes to communicate with node B, e.g. to send a
+FINDNODE query.
+
+Node A must have a node record for node B and know B's node ID to communicate with it. If
+node A has session keys from prior communication, it encrypts its request with those keys.
+If no keys are known, it initiates the handshake by sending a packet with random content.
+
+    A -> B   FINDNODE encrypted with unknown key or random-packet
+
+Node B receives the initial packet, extracts the source node ID from the packet (see
+[encoding section]), and continues the handshake by responding with [WHOAREYOU]. The
+WHOAREYOU packet contains the id-nonce value to be signed by A as well as the highest
+known ENR sequence number of node A's record.
+
+    A <- B   WHOAREYOU including id-nonce, enr-seq
+
+Node A now knows that node B is alive and is ready to perform the handshake. The handshake
+proceeds by re-sending the original request message in a [handshake packet].
+
+The handshake packet includes an ephemeral public key in the cryptosystem used by B's
+identity scheme (e.g. an elliptic curve key on the secp256k1 curve if node B uses the "v4"
+scheme). This key is used to perform Diffie-Hellman key agreement with B's static public
+key and the session keys are derived from it using the HKDF key derivation function.
+
+    ephemeral-key    = random private key
+    ephemeral-pubkey = public key corresponding to ephemeral-key
+    dest-pubkey      = public key of B
+    secret           = ecdh(ephemeral-key, dest-pubkey)
+    info             = "discovery v5 key agreement" || node-id-A || node-id-B
+    prk              = HKDF-Extract(secret, id-nonce)
+
+    initiator-key, recipient-key, auth-resp-key = HKDF-Expand(prk, info)
```

Comment:
Seems like we don't need `auth-resp-key` anymore.

Reply:
Yes, that's correct. Good catch!